tfhe/src/integer/server_key/radix/scalar_sub.rs

```diff
@@ -34,44 +36,70 @@ impl ServerKey {
     /// let dec: u64 = cks.decrypt(&ct_res);
     /// assert_eq!(msg - scalar, dec);
     /// ```
-    pub fn unchecked_scalar_sub<PBSOrder: PBSOrderMarker>(
+    pub fn unchecked_scalar_sub<PBSOrder, T>(
         &self,
         ct: &RadixCiphertext<PBSOrder>,
-        scalar: u64,
-    ) -> RadixCiphertext<PBSOrder> {
+        scalar: T,
+    ) -> RadixCiphertext<PBSOrder>
+    where
+        PBSOrder: PBSOrderMarker,
+        T: UnsignedInteger + DecomposableInto<u8>,
+    {
         let mut result = ct.clone();
         self.unchecked_scalar_sub_assign(&mut result, scalar);
         result
     }
 
-    pub fn unchecked_scalar_sub_assign<PBSOrder: PBSOrderMarker>(
-        &self,
-        ct: &mut RadixCiphertext<PBSOrder>,
-        scalar: u64,
-    ) {
-        // Widen to 128 bits to avoid overflow when doing wrapped negation.
-        let scalar = scalar as u128;
-        //Bits of message put to 1
-        let mask = (self.key.message_modulus.0 - 1) as u128;
+    fn create_negated_block_decomposer<T>(&self, scalar: T) -> impl Iterator<Item = u8>
+    where
+        T: UnsignedInteger + DecomposableInto<u8>,
+    {
+        let bits_in_message = self.key.message_modulus.0.ilog2();
+        assert!(bits_in_message <= u8::BITS);
 
-        let modulus = (self.key.message_modulus.0 as u128).checked_pow(ct.blocks.len() as u32);
+        // The whole idea behind this iterator we construct is:
+        // - to support combos of parameters and num blocks for which the total number of bits is
+        //   not a multiple of T::BITS
+        //
+        // - Support subtraction in the case the T::BITS is lower than the target ciphertext bits.
+        //   In clear rust this would require an upcast, to support that we have to do a few things
 
-        let neg_scalar = match modulus {
-            Some(modul) => scalar.wrapping_neg() % modul,
-            None => scalar.wrapping_neg(),
-        };
+        //  The negation of a numeric value of type T, in two complemement
+        //  is neg(v) = bitwise_not(v) + 1
+        let flipped = !scalar;
+        let (neg_scalar, has_overflowed) = flipped.overflowing_add(T::ONE);
 
-        let mut power = 1_u128;
-        //Put each decomposition into a new ciphertext
-        for ct_i in ct.blocks.iter_mut() {
-            let mut decomp = neg_scalar & (mask * power);
-            decomp /= power;
+        // If we had upcasted the scalar, its msb would be zeros (0)
+        // then they would become ones (1) after the bitwise_not (!).
+        // We need to know if after the `+1` these padded bits are 0s or 1s.
+        // And we get this information by looking at if addition overflowed
+        // meaning there is a carry.
+        let padding_bit = if has_overflowed { 0u32 } else { 1u32 };
```

Comment:
the only case this can happen is for the value 0 right ? do we need to handle that case ? I'm not sure quite yet

Reply:
Yes I think its the only case it can happen, we could use `  let padding_bit = if scalar == T::ZERO { 0u32 } else { 1u32 };
` or `let padding_bit = u32::from(scalar != T::ZERO)`

Reply:
my question is more do we have to have a special case for the padding bits in case the scalar is 0 ?

Reply:
Well, This handles the special case of when the scalar is zero, but its less obvious the way it is currently written

Reply:
if scalar is zero we could return an empty iterator as subtractring 0 means subtractring nothing

Reply:
agreed, I think skipping computations for scalar == 0 is ok

Reply:
Opted for returning an Option<impl Iterator> and handling that case in the 2 function that needed to handle it